rfd/0163-vnet.md

```diff
@@ -0,0 +1,813 @@
+  rpc GetVNet(GetVNetRequest) returns (VNet);
+
+  // CreateVNet creates a new VNet resource.
+  rpc CreateVNet(CreateVNetRequest) returns (VNet);
+
+  // UpdateVNet updates an existing VNet resource.
+  rpc UpdateVNet(UpdateVNetRequest) returns (VNet);
+
+  // DeleteVNet hard deletes the specified VNet resource.
+  rpc DeleteVNet(DeleteVNetRequest) returns (google.protobuf.Empty);
+}
+
+// Request for GetVNet.
+message GetVNetRequest {}
+
+// Request for CreateVNet.
+message CreateVNetRequest {
+  // The VNet resource to create.
+  VNet vnet = 1;
+}
+
+// Request for UpdateVNet.
+message UpdateVNetRequest {
+  // The VNet resource to create.
+  VNet vnet = 1;
+
+  // The update mask applied to a VNet.
+  // Fields are masked according to their proto name.
+  FieldMask update_mask = 2;
+}
+
+// Request for DeleteVNet.
+message DeleteVNetRequest {}
+```
+
+The current `AppSpecV3` message will get a new field to hold the custom VNet
+address for apps.
+
+```
+// AppSpecV3 is the AppV3 resource spec.
+message AppSpecV3 {
+  // VnetAddr is the VNet address the application is accessible at.
+  string vnet_addr = 9;
+}
+```
+
+### Backward Compatibility
+
+We will not be changing or interfering with existing app access to backward
+compatibility is not an issue.
+
+### Audit Events
+
+VNet is a client-side feature so no audit events will be added.
+
+### Observability
+
+VNet is a client-side feature there will be no new logs, audit events, or
+metrics server-side.
+Usage will be observable via telemetry, see [Product Usage](#product-usage).
+
+Client-side, connections and errors will be visible in the UI in the VNet tab,
+see [Managing VNet](#managing-vnet).
+
+
+### Product Usage
+
+As tsh tracks usage only indirectly through the auth server and VNet is mostly a client-side
+feature, the only readily available option is to track usage through Connect's telemetry.
+
+Connect already tracks access to applications through [the `connect.protocol.use`
+event](https://github.com/gravitational/teleport/blob/5688a13f71f1fb011fa0b884a304bedf325b7c1a/proto/prehog/v1alpha/connect.proto#L38-L47).
+It counts each browser launch and each setup of a local proxy.
+
+In order to compare usage through VNet vs other means, we send `connect.protocol.use` each time a
+successful connection is made through VNet. To differentiate between VNet and non-VNet usage, we add
+a new string field to `connect.protocol.use` called `access_method`. It contains one of the three
+supported values:
+
+- `proxy_service`: a resource was accessed through the proxy service. In the case of apps, it means
+  launching an app in the browser through the `/web/launch` URL of the proxy service.
+- `local_proxy`: a local proxy was set up to access a resource.
+  - As of today, Connect sends only one event per local proxy when a local proxy gets created, not
+    for each new connection established through the proxy.
+- `vnet`: a new connection to an app was made through VNet.
```

Comment:
This distinction makes sense to me, although I'm not sure about the `proxy_service` value. I have a feeling it may be difficult to recall what it means when looking on a dashboard. Perhaps something like `one_time_run` would be more descriptive?
But on the other hand, it doesn't align very well with the rest of the values, so maybe `proxy_service` is fine.

Reply:
I rewrote the section on Product Usage a little bit. We no longer send one event per connection, instead we send just one `connect.protocol.use` event per app per VNet lifespan.

It's even more straightforward to implement than sending one event per connection, since for that scenario we'd have to update our local proxy app middleware to send those events.

Now that [we have rethought how to measure protocol popularity](https://gravitational.slack.com/archives/C03FJA391M3/p1709038088363419?thread_ts=1708966249.698449&cid=C03FJA391M3), sending one event per app still properly tracks app usage. And since it mimics how local proxies send events (just one event when a local proxy is created), it'll allow us to compare usage of VNet vs local proxies for apps.

---

I also changed the name of the field from `access_method` to `access_through`. I feel it sounds more natural. As for `proxy_service`, I don't have a better idea. As you said, considering the other two values, it makes sense.

I was also thinking of `"access_method": "direct"` but this might give the wrong idea. OTOH, in all three cases the resource ends up getting accessed through the proxy service.